modules/samples/firealarm/component/manager/src/main/java/org/homeautomation/firealarm/manager/api/ManagerService.java

```diff
@@ -0,0 +1,337 @@
+            device.setName(name);
+            device.setType(DeviceTypeConstants.DEVICE_TYPE);
+            boolean updated = APIUtil.getDeviceManagementService().modifyEnrollment(device);
+            if (updated) {
+                response.setStatus(Response.Status.OK.getStatusCode());
+            } else {
+                response.setStatus(Response.Status.NOT_ACCEPTABLE.getStatusCode());
+            }
+            return updated;
+        } catch (DeviceManagementException e) {
+            log.error(e.getErrorMessage());
+            return false;
+        }
+    }
+
+    /**
+     * Get device information
+     *
+     * @param deviceId unique identifier for device
+     * @return device
+     */
+    @Path("manager/device/{device_id}")
+    @GET
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    public Device getDevice(@PathParam("device_id") String deviceId) {
+        DeviceIdentifier deviceIdentifier = new DeviceIdentifier();
+        deviceIdentifier.setId(deviceId);
+        deviceIdentifier.setType(DeviceTypeConstants.DEVICE_TYPE);
+        try {
+            return APIUtil.getDeviceManagementService().getDevice(deviceIdentifier);
+        } catch (DeviceManagementException ex) {
+            log.error("Error occurred while retrieving device with Id " + deviceId + "\n" + ex);
+            return null;
+        }
+    }
+
+    /**
+     * This will download the agent for given device type
+     *
+     * @param deviceName name of the device which is to be created
+     * @param sketchType name of sketch type
+     * @return agent archive
+     */
+    @Path("manager/device/{sketch_type}/download")
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response downloadSketch(@QueryParam("deviceName") String deviceName,
+                                   @PathParam("sketch_type") String sketchType) {
+        try {
+            ZipArchive zipFile = createDownloadFile(APIUtil.getAuthenticatedUser(), deviceName, sketchType);
+            Response.ResponseBuilder response = Response.ok(FileUtils.readFileToByteArray(zipFile.getZipFile()));
+            response.type("application/zip");
+            response.header("Content-Disposition", "attachment; filename=\"" + zipFile.getFileName() + "\"");
+            return response.build();
+        } catch (IllegalArgumentException ex) {
+            return Response.status(400).entity(ex.getMessage()).build();//bad request
+        } catch (DeviceManagementException ex) {
+            return Response.status(500).entity(ex.getMessage()).build();
+        } catch (AccessTokenException ex) {
+            return Response.status(500).entity(ex.getMessage()).build();
+        } catch (DeviceControllerException ex) {
+            return Response.status(500).entity(ex.getMessage()).build();
+        } catch (IOException ex) {
+            return Response.status(500).entity(ex.getMessage()).build();
```

Comment:
it's better to log the error at server-side in case of an exception (should happen only at API level, 
otherwise it would be redundant), so that it would be a good piece of information for system administrators to debug issues.

When logging an error, it should be something custom and specific to API layer as the exception actually occurs due to a call at that level. And also, it's better to pass the error stack-trace to be logged, so that errors and its causes could be traced so easily at the event of an exception. This can be done simply by passing the exception object being thrown, to log.error function as a parameter.

log.error(errorMessage, e);